rhasspywake_raven/dtw.py

```diff
@@ -90,32 +81,26 @@ def _compute_optimal_path(
         if len(y.shape) == 1:
             y = y.reshape(-1, 1)
 
-        distance_matrix = scipy.spatial.distance.cdist(x, y, metric=self.distance_func)
-
-        cost_matrix = np.full(shape=(m, n), fill_value=math.inf, dtype=float)
-        cost_matrix[0][0] = distance_matrix[0][0]
+        cost_matrix = scipy.spatial.distance.cdist(x, y, metric=self.distance_func)
 
         for row in range(1, m):
-            cost = distance_matrix[row, 0]
-            cost_matrix[row][0] = cost + cost_matrix[row - 1][0]
+            cost_matrix[row, 0] += cost_matrix[row - 1, 0]
 
         for col in range(1, n):
-            cost = distance_matrix[0, col]
-            cost_matrix[0][col] = cost + cost_matrix[0][col - 1]
+            cost_matrix[0, col] += cost_matrix[0, col - 1]
 
         for row in range(1, m):
             for col in range(1, n):
-                cost = distance_matrix[row, col]
-                cost_matrix[row][col] = cost + min(
-                    cost_matrix[row - 1][col],  # insertion
-                    cost_matrix[row][col - 1],  # deletion
-                    cost_matrix[row - 1][col - 1],  # match
+                cost_matrix[row, col] += min(
+                    cost_matrix[row - 1, col],  # insertion
+                    cost_matrix[row, col - 1],  # deletion
+                    cost_matrix[row - 1, col - 1],  # match
```

Comment:
This reduces memory allocations by 50%, since we do not really need two matrices. This is done in a similar way in the javascript implementation, where the similarity is calculated while iterating:
https://github.com/mathquis/node-personal-wakeword/blob/ba6375edae78c7733b23bd7a73476a671fc89f52/src/dtw/dtw.js#L44

_compute_optimal_path_with_window could even save 2/3 of the memory allocations. Both the new cost_matrix and the copy into the results:
```
self.cost_matrix = cost_matrix[1:, 1:]
```
could be saved. However this change is more convoluted (it requires the iteration to be split into multiple parts to account for the missing first row/column)